src/openms/source/ANALYSIS/ID/SimpleSearchEngineAlgorithm.cpp

```diff
@@ -0,0 +1,654 @@
+    {
+      cout << "duplicate variable modification provided." << endl;
+      return ExitCodes::ILLEGAL_PARAMETERS;
+    }
+
+    vector<ResidueModification> fixed_modifications = getModifications_(modifications_fixed_);
+    vector<ResidueModification> variable_modifications = getModifications_(modifications_variable_);
+
+    // load MS2 map
+    PeakMap spectra;
+    MzMLFile f;
+    //f.setLogType(log_type_);
+
+    PeakFileOptions options;
+    options.clearMSLevels();
+    options.addMSLevel(2);
+    f.getOptions() = options;
+    f.load(in_mzML, spectra);
+    spectra.sortSpectra(true);
+
+    startProgress(0, 1, "Filtering spectra...");
+    preprocessSpectra_(spectra, fragment_mass_tolerance_, fragment_mass_tolerance_unit_ppm);
+    endProgress();
+
+    // build multimap of precursor mass to scan index
+    multimap<double, Size> multimap_mass_2_scan_index;
+    for (PeakMap::ConstIterator s_it = spectra.begin(); s_it != spectra.end(); ++s_it)
+    {
+      int scan_index = s_it - spectra.begin();
+      vector<Precursor> precursor = s_it->getPrecursors();
+
+      // there should only one precursor and MS2 should contain at least a few peaks to be considered (e.g. at least for every AA in the peptide)
+      if (precursor.size() == 1 && s_it->size() >= peptide_min_size_)
+      {
+        Size precursor_charge = precursor[0].getCharge();
+
+        if (precursor_charge < precursor_min_charge_ 
+         || precursor_charge > precursor_max_charge_)
+        {
+          continue;
+        }
+
+        double precursor_mz = precursor[0].getMZ();
+
+        // calculate precursor mass (optionally corrected for misassignment) and map it to MS scan index
+        for (int isotope_number : precursor_isotopes_)
+        {
+          double precursor_mass = (double) precursor_charge * precursor_mz - (double) precursor_charge * Constants::PROTON_MASS_U;
+
+          // correct for monoisotopic misassignments of the precursor annotation
+          if (isotope_number != 0) { precursor_mass -= isotope_number * Constants::C13C12_MASSDIFF_U; }
+
+          multimap_mass_2_scan_index.insert(make_pair(precursor_mass, scan_index));
+        }
+      }
+    }
+
+    // create spectrum generator
+    TheoreticalSpectrumGenerator spectrum_generator;
+    Param param(spectrum_generator.getParameters());
+    param.setValue("add_first_prefix_ion", "true");
+    param.setValue("add_metainfo", "true");
+    spectrum_generator.setParameters(param);
+
+    // preallocate storage for PSMs
+    vector<vector<AnnotatedHit_> > annotated_hits(spectra.size(), vector<AnnotatedHit_>());
+    for (auto & a : annotated_hits) { a.reserve(2 * report_top_hits_); }
+
+#ifdef _OPENMP
+    // we want to do locking at the spectrum level so we get good parallelisation 
+    vector<omp_lock_t> annotated_hits_lock(annotated_hits.size());
+    for (size_t i = 0; i != annotated_hits_lock.size(); i++) { omp_init_lock(&(annotated_hits_lock[i])); }
```

Comment:
is there a performance issue with using many locks in OpenMP? These will be potentially hundreds of thousands of locks...

Reply:
Good question and I don't know the definite answer to that. For sure, it scales much better if every single spectrum is locked. Also, a critical section is internally also modelled using a lock. As the locks are in the shared memory there might be some speed problems related to loading the locks into the cache. @cbielow what is your take on that?